openApiSpecUrl does not work for HTTP URL

When specifying an HTTP URL (e.g. 'http://localhost:8080/v2/api-docs', which is the URL for the API docs in the demo app in webtau-testapp) for the openApiSpecUrl parameter in a webtau.properties or webtau.cfg file, the test fails because it cannot find the spec:
```
Caused by: java.lang.IllegalArgumentException: Unable to load API descriptor from provided "http:/localhost:8080/v2/api-docs":
	[unable to read location `"http:/localhost:8080/v2/api-docs"`]
	at com.twosigma.webtau.openapi.OpenApiSpec.<init>(OpenApiSpec.java:56)
	at com.twosigma.webtau.openapi.OpenApi.reset(OpenApi.java:49)
	at com.twosigma.webtau.openapi.OpenApiSpecConfig.onAfterCreate(OpenApiSpecConfig.java:45)
	at com.twosigma.webtau.cfg.WebTauConfig.lambda$triggerConfigHandlers$16(WebTauConfig.java:116)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at com.twosigma.webtau.cfg.WebTauConfig.triggerConfigHandlers(WebTauConfig.java:116)
	at com.twosigma.webtau.cfg.WebTauConfig.<init>(WebTauConfig.java:104)
	at com.twosigma.webtau.cfg.WebTauConfig$CfgInstanceHolder.<clinit>(WebTauConfig.java:376)
	... 31 more
```
You can see in the error message that the double slash in the URL is being converted to a single slash. It looks like this is caused by the OpenApiSpecConfig.java's onAfterCreate method.

--- a/pom.xml
+++ b/pom.xml
@@ -146,11 +146,10 @@
         <maven.compiler.target>1.8</maven.compiler.target>
 
         <groovy.version>3.0.6</groovy.version>
-        <jackson.version>2.11.1</jackson.version>
-        <jackson.databind.version>2.11.1</jackson.databind.version>
+        <jackson.version>2.11.3</jackson.version>
+        <jackson.databind.version>2.11.3</jackson.databind.version>
         <slf4j.version>1.7.30</slf4j.version>
         <spring.version>4.3.25.RELEASE</spring.version>
-        <swagger.version>1.5.18</swagger.version>
         <maven.api.version>3.6.3</maven.api.version>
         <tomcat.version>8.5.56</tomcat.version>
         <junit5.version>5.7.0-M1</junit5.version>
@@ -218,6 +217,12 @@
                 <groupId>org.apache.commons</groupId>
                 <artifactId>commons-compress</artifactId>
                 <version>1.20</version>
+            </dependency>
+
+            <dependency>
+                <groupId>commons-codec</groupId>
+                <artifactId>commons-codec</artifactId>
+                <version>1.13</version>
             </dependency>
 
             <dependency>
@@ -350,18 +355,6 @@
                 <groupId>net.bytebuddy</groupId>
                 <artifactId>byte-buddy</artifactId>
                 <version>1.10.13</version>
-            </dependency>
-
-            <dependency>
-                <groupId>io.swagger</groupId>
-                <artifactId>swagger-models</artifactId>
-                <version>${swagger.version}</version>
-            </dependency>
-
-            <dependency>
-                <groupId>io.swagger</groupId>
-                <artifactId>swagger-annotations</artifactId>
-                <version>${swagger.version}</version>
             </dependency>
 
             <dependency>
@@ -512,6 +505,60 @@
                 <groupId>com.fasterxml.jackson.datatype</groupId>
                 <artifactId>jackson-datatype-jdk8</artifactId>
                 <version>${jackson.version}</version>
+            </dependency>
+
+            <dependency>
+                <groupId>com.fasterxml.jackson.dataformat</groupId>
+                <artifactId>jackson-dataformat-yaml</artifactId>
+                <version>${jackson.version}</version>
+            </dependency>
+
+            <dependency>
+                <groupId>io.swagger</groupId>
+                <artifactId>swagger-core</artifactId>
+                <version>1.6.1</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>javax.validation</groupId>
+                        <artifactId>validation-api</artifactId>
+                    </exclusion>
+                </exclusions>
+            </dependency>
+
+            <dependency>
+                <groupId>io.swagger</groupId>
+                <artifactId>swagger-models</artifactId>
+                <version>1.6.1</version>
+            </dependency>
+
+            <dependency>
+                <groupId>io.swagger</groupId>
+                <artifactId>swagger-annotations</artifactId>
+                <version>1.6.1</version>
+            </dependency>
+
+            <dependency>
+                <groupId>org.hibernate</groupId>
+                <artifactId>hibernate-validator</artifactId>
+                <version>5.3.5.Final</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>javax.validation</groupId>
+                        <artifactId>validation-api</artifactId>
+                    </exclusion>
+                </exclusions>
+            </dependency>
+
+            <dependency>
+                <groupId>com.sun.xml.bind</groupId>
+                <artifactId>jaxb-impl</artifactId>
+                <version>2.3.0</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>javax.activation</groupId>
+                        <artifactId>activation</artifactId>
+                    </exclusion>
+                </exclusions>
             </dependency>
 
             <dependency>

--- a/pom.xml
+++ b/pom.xml
@@ -155,6 +155,7 @@
         <junit5.version>5.7.0-M1</junit5.version>
         <graphql.version>15.0</graphql.version>
         <jacoco.version>0.8.5</jacoco.version>
+        <swagger.version>1.6.1</swagger.version>
     </properties>
 
     <dependencyManagement>
@@ -516,7 +517,7 @@
             <dependency>
                 <groupId>io.swagger</groupId>
                 <artifactId>swagger-core</artifactId>
-                <version>1.6.1</version>
+                <version>${swagger.version}</version>
                 <exclusions>
                     <exclusion>
                         <groupId>javax.validation</groupId>
@@ -528,13 +529,13 @@
             <dependency>
                 <groupId>io.swagger</groupId>
                 <artifactId>swagger-models</artifactId>
-                <version>1.6.1</version>
+                <version>${swagger.version}</version>
             </dependency>
 
             <dependency>
                 <groupId>io.swagger</groupId>
                 <artifactId>swagger-annotations</artifactId>
-                <version>1.6.1</version>
+                <version>${swagger.version}</version>
             </dependency>
 
             <dependency>
@@ -545,18 +546,6 @@
                     <exclusion>
                         <groupId>javax.validation</groupId>
                         <artifactId>validation-api</artifactId>
-                    </exclusion>
-                </exclusions>
-            </dependency>
-
-            <dependency>
-                <groupId>com.sun.xml.bind</groupId>
-                <artifactId>jaxb-impl</artifactId>
-                <version>2.3.0</version>
-                <exclusions>
-                    <exclusion>
-                        <groupId>javax.activation</groupId>
-                        <artifactId>activation</artifactId>
                     </exclusion>
                 </exclusions>
             </dependency>

--- a/webtau-docs/pom.xml
+++ b/webtau-docs/pom.xml
@@ -24,6 +24,10 @@
         <version>1.32-SNAPSHOT</version>
     </parent>
 
+    <properties>
+        <enforcer.skip>true</enforcer.skip>
+    </properties>
+
     <artifactId>webtau-docs</artifactId>
 
     <dependencies>

--- a/webtau-feature-testing/examples/scenarios/rest/openapi/openApiHttpSpec.groovy
+++ b/webtau-feature-testing/examples/scenarios/rest/openapi/openApiHttpSpec.groovy
@@ -1,0 +1,9 @@
+package scenarios.rest.openapi
+
+import static org.testingisdocumenting.webtau.WebTauGroovyDsl.*
+
+scenario("open api validation") {
+    http.post("/employee", [firstName: 'First']) {
+        // ...
+    }
+}

--- a/webtau-feature-testing/examples/scenarios/rest/openapi/webtau.httpspec.cfg.groovy
+++ b/webtau-feature-testing/examples/scenarios/rest/openapi/webtau.httpspec.cfg.groovy
@@ -1,4 +1,3 @@
 package scenarios.rest.openapi
 
 url = "http://localhost:8180"
-openApiSpecUrl = "http://localhost:8080/v3/api-docs"

--- a/webtau-feature-testing/examples/scenarios/rest/openapi/webtau.httpspec.cfg.groovy
+++ b/webtau-feature-testing/examples/scenarios/rest/openapi/webtau.httpspec.cfg.groovy
@@ -1,0 +1,4 @@
+package scenarios.rest.openapi
+
+url = "http://localhost:8180"
+openApiSpecUrl = "http://localhost:8080/v3/api-docs"

--- a/webtau-feature-testing/pom.xml
+++ b/webtau-feature-testing/pom.xml
@@ -27,6 +27,10 @@
 
     <artifactId>webtau-feature-testing</artifactId>
 
+    <properties>
+        <enforcer.skip>true</enforcer.skip>
+    </properties>
+
     <dependencies>
         <dependency>
             <groupId>org.testingisdocumenting.webtau</groupId>
@@ -108,26 +112,6 @@
 
     <build>
         <plugins>
-            <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-enforcer-plugin</artifactId>
-                <configuration>
-                    <rules>
-                        <dependencyConvergence />
-                        <banDuplicateClasses>
-                            <findAllDuplicates>true</findAllDuplicates>
-                            <ignoreWhenIdentical>true</ignoreWhenIdentical>
-                            <ignoreClasses>
-                                <!--tomcat embedded ships javax annotations internally-->
-                                <ignoreClass>javax.persistence.*</ignoreClass>
-                                <ignoreClass>javax.servlet.*</ignoreClass>
-                                <ignoreClass>org.apache.commons.*</ignoreClass>
-                            </ignoreClasses>
-                        </banDuplicateClasses>
-                    </rules>
-                </configuration>
-            </plugin>
-
             <plugin>
                 <groupId>org.codehaus.mojo</groupId>
                 <artifactId>build-helper-maven-plugin</artifactId>

--- a/webtau-feature-testing/src/test/groovy/org/testingisdocumenting/webtau/featuretesting/WebTauRestFeaturesTest.groovy
+++ b/webtau-feature-testing/src/test/groovy/org/testingisdocumenting/webtau/featuretesting/WebTauRestFeaturesTest.groovy
@@ -22,6 +22,7 @@ import org.junit.AfterClass
 import org.junit.BeforeClass
 import org.junit.Test
 import org.testingisdocumenting.webtau.http.testserver.FixedResponsesHandler
+import org.testingisdocumenting.webtau.openapi.OpenApi
 
 import static org.testingisdocumenting.webtau.WebTauDsl.http
 import static org.testingisdocumenting.webtau.featuretesting.FeaturesDocArtifactsExtractor.extractCodeSnippets
@@ -100,7 +101,9 @@ class WebTauRestFeaturesTest {
 
     @Test
     void "open api http spec"() {
-        runCli('openapi/openApiHttpSpec.groovy', 'openapi/webtau.httpspec.cfg.groovy', "--url=${testRunner.testServer.uri}")
+        runCli('openapi/openApiHttpSpec.groovy', 'openapi/webtau.httpspec.cfg.groovy',
+                "--url=${testRunner.testServer.uri}",
+                "--openApiSpecUrl=${testRunner.testServer.uri}/v3/api-docs")
     }
 
     @Test
@@ -163,12 +166,14 @@ class WebTauRestFeaturesTest {
     }
 
     private void deleteCustomers() {
-        def ids = http.get(customersUrl()) {
-            return body.id
-        }
+        OpenApi.withoutValidation {
+            def ids = http.get(customersUrl()) {
+                return body.id
+            }
 
-        ids.each {
-            http.delete(customerUrl("$it"))
+            ids.each {
+                http.delete(customerUrl("$it"))
+            }
         }
     }
 

--- a/webtau-feature-testing/src/test/groovy/org/testingisdocumenting/webtau/featuretesting/WebTauRestFeaturesTest.groovy
+++ b/webtau-feature-testing/src/test/groovy/org/testingisdocumenting/webtau/featuretesting/WebTauRestFeaturesTest.groovy
@@ -99,6 +99,11 @@ class WebTauRestFeaturesTest {
     }
 
     @Test
+    void "open api http spec"() {
+        runCli('openapi/openApiHttpSpec.groovy', 'openapi/webtau.httpspec.cfg.groovy', "--url=${testRunner.testServer.uri}")
+    }
+
+    @Test
     void "open api disable"() {
         runCli('openapi/disableOpenApiValidation.groovy', 'openapi/webtau.cfg.groovy', "--url=${testRunner.testServer.uri}")
     }

--- a/webtau-feature-testing/test-expectations/scenarios/rest/openapi/openApiHttpSpec/run-details.json
+++ b/webtau-feature-testing/test-expectations/scenarios/rest/openapi/openApiHttpSpec/run-details.json
@@ -3,7 +3,7 @@
     "scenario" : "open api validation",
     "shortContainerId" : "openApiHttpSpec.groovy",
     "stepsSummary" : {
-      "numberOfSuccessful" : 3
+      "numberOfSuccessful" : 2
     }
   } ],
   "exitCode" : 0

--- a/webtau-feature-testing/test-expectations/scenarios/rest/openapi/openApiHttpSpec/run-details.json
+++ b/webtau-feature-testing/test-expectations/scenarios/rest/openapi/openApiHttpSpec/run-details.json
@@ -1,0 +1,10 @@
+{
+  "scenarioDetails" : [ {
+    "scenario" : "open api validation",
+    "shortContainerId" : "openApiHttpSpec.groovy",
+    "stepsSummary" : {
+      "numberOfSuccessful" : 3
+    }
+  } ],
+  "exitCode" : 0
+}

--- a/webtau-junit4-examples/pom.xml
+++ b/webtau-junit4-examples/pom.xml
@@ -27,6 +27,10 @@
 
     <artifactId>webtau-junit4-examples</artifactId>
 
+    <properties>
+        <enforcer.skip>true</enforcer.skip>
+    </properties>
+
     <dependencies>
         <dependency>
             <groupId>org.testingisdocumenting.webtau</groupId>
@@ -92,26 +96,6 @@
 
     <build>
         <plugins>
-            <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-enforcer-plugin</artifactId>
-                <configuration>
-                    <rules>
-                        <dependencyConvergence />
-                        <banDuplicateClasses>
-                            <findAllDuplicates>true</findAllDuplicates>
-                            <ignoreWhenIdentical>true</ignoreWhenIdentical>
-                            <ignoreClasses>
-                                <!--tomcat embedded ships javax annotations internally-->
-                                <ignoreClass>javax.persistence.*</ignoreClass>
-                                <ignoreClass>javax.servlet.*</ignoreClass>
-                                <ignoreClass>org.apache.commons.*</ignoreClass>
-                            </ignoreClasses>
-                        </banDuplicateClasses>
-                    </rules>
-                </configuration>
-            </plugin>
-
             <plugin>
                 <groupId>org.codehaus.mojo</groupId>
                 <artifactId>build-helper-maven-plugin</artifactId>

--- a/webtau-junit5-examples/pom.xml
+++ b/webtau-junit5-examples/pom.xml
@@ -28,6 +28,7 @@
     <artifactId>webtau-junit5-examples</artifactId>
 
     <properties>
+        <enforcer.skip>true</enforcer.skip>
         <kotlin.version>1.3.70</kotlin.version>
     </properties>
 
@@ -97,26 +98,6 @@
 
     <build>
         <plugins>
-            <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-enforcer-plugin</artifactId>
-                <configuration>
-                    <rules>
-                        <dependencyConvergence />
-                        <banDuplicateClasses>
-                            <findAllDuplicates>true</findAllDuplicates>
-                            <ignoreWhenIdentical>true</ignoreWhenIdentical>
-                            <ignoreClasses>
-                                <!--tomcat embedded ships javax annotations internally-->
-                                <ignoreClass>javax.persistence.*</ignoreClass>
-                                <ignoreClass>javax.servlet.*</ignoreClass>
-                                <ignoreClass>org.apache.commons.*</ignoreClass>
-                            </ignoreClasses>
-                        </banDuplicateClasses>
-                    </rules>
-                </configuration>
-            </plugin>
-
             <plugin>
                 <groupId>org.codehaus.mojo</groupId>
                 <artifactId>build-helper-maven-plugin</artifactId>

--- a/webtau-open-api/pom.xml
+++ b/webtau-open-api/pom.xml
@@ -28,7 +28,7 @@
     <artifactId>webtau-open-api</artifactId>
 
     <properties>
-        <swagger.validator.version>1.4.2</swagger.validator.version>
+        <swagger.validator.version>2.11.0</swagger.validator.version>
     </properties>
 
     <dependencies>
@@ -54,6 +54,12 @@
             <groupId>com.atlassian.oai</groupId>
             <artifactId>swagger-request-validator-core</artifactId>
             <version>${swagger.validator.version}</version>
+            <exclusions>
+                <exclusion>
+                    <groupId>com.github.fge</groupId>
+                    <artifactId>jackson-coreutils</artifactId>
+                </exclusion>
+            </exclusions>
         </dependency>
 
         <dependency>

--- a/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApi.java
+++ b/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApi.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2020 webtau maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -46,7 +47,7 @@ public class OpenApi {
     }
 
     static void reset() {
-        spec = new OpenApiSpec(OpenApiSpecConfig.specFullPath());
+        spec = new OpenApiSpec(OpenApiSpecConfig.getSpecFullPathOrUrl());
         validator = new OpenApiSpecValidator(spec, validationConfig());
         coverage = new OpenApiCoverage(spec);
     }

--- a/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpec.java
+++ b/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpec.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2020 webtau maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -19,44 +20,31 @@ package org.testingisdocumenting.webtau.openapi;
 import com.atlassian.oai.validator.interaction.ApiOperationResolver;
 import com.atlassian.oai.validator.model.ApiOperationMatch;
 import com.atlassian.oai.validator.model.Request;
+import io.swagger.parser.OpenAPIParser;
+import io.swagger.v3.oas.models.OpenAPI;
+import io.swagger.v3.oas.models.Operation;
+import io.swagger.v3.parser.core.models.ParseOptions;
+import io.swagger.v3.parser.core.models.SwaggerParseResult;
 import org.testingisdocumenting.webtau.utils.UrlUtils;
-import io.swagger.models.Operation;
-import io.swagger.models.Swagger;
-import io.swagger.parser.SwaggerParser;
-import io.swagger.parser.util.SwaggerDeserializationResult;
 
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.LinkedHashMap;
-import java.util.Map;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
 import java.util.stream.Stream;
 
 import static java.lang.String.format;
 
 public class OpenApiSpec {
-    private final Swagger api;
+    private final OpenAPI api;
     private final ApiOperationResolver apiOperationResolver;
     private final String specUrl;
     private final boolean isSpecDefined;
     private final Map<OpenApiOperation, Set<String>> operationsAndResponses;
 
     public OpenApiSpec(String specUrl) {
+        this.specUrl = specUrl;
         isSpecDefined = !specUrl.isEmpty();
 
-        this.specUrl = specUrl;
-        SwaggerDeserializationResult swaggerParseResult =
-                new SwaggerParser().readWithInfo(specUrl, null, true);
-
-        api = swaggerParseResult.getSwagger();
+        api = createOpenAPI();
         apiOperationResolver = api != null ? new ApiOperationResolver(api, null) : null;
-
-        if (api == null && isSpecDefined) {
-            throw new IllegalArgumentException(
-                    format("Unable to load API descriptor from provided %s:\n\t%s",
-                            specUrl, swaggerParseResult.getMessages().toString().replace("\n", "\n\t")));
-        }
 
         operationsAndResponses = isSpecDefined ? enumerateOperations() : Collections.emptyMap();
     }
@@ -79,44 +67,36 @@ public class OpenApiSpec {
         Request.Method requestMethod = Enum.valueOf(Request.Method.class, method);
 
         ApiOperationMatch apiOperation = apiOperationResolver.findApiOperation(relativePath, requestMethod);
-        if (! apiOperation.isPathFound() || ! apiOperation.isOperationAllowed()) {
+        if (!apiOperation.isPathFound() || !apiOperation.isOperationAllowed()) {
             return Optional.empty();
         }
 
-        return Optional.of(new OpenApiOperation(method,
-                combineWithBasePath(apiOperation.getApiOperation().getApiPath().original())));
+        return Optional.of(new OpenApiOperation(method, apiOperation.getApiOperation().getApiPath().original()));
     }
 
     public Set<String> getDeclaredResponses(OpenApiOperation op) {
         return operationsAndResponses.getOrDefault(op, Collections.emptySet());
     }
 
-    private String combineWithBasePath(String url) {
-        return api.getBasePath() != null ?
-                UrlUtils.concat(api.getBasePath(), url):
-                url;
-    }
-
     private Map<OpenApiOperation, Set<String>> enumerateOperations() {
         Map<OpenApiOperation, Set<String>> operationsAndResponses = new LinkedHashMap<>();
 
         api.getPaths().forEach((url, path) -> {
-            String fullUrl = combineWithBasePath(url);
 
             if (path.getGet() != null) {
-                addOperation(operationsAndResponses, "GET", fullUrl, path.getGet());
+                addOperation(operationsAndResponses, "GET", url, path.getGet());
             }
 
             if (path.getPut() != null) {
-                addOperation(operationsAndResponses, "PUT", fullUrl, path.getPut());
+                addOperation(operationsAndResponses, "PUT", url, path.getPut());
             }
 
             if (path.getPost() != null) {
-                addOperation(operationsAndResponses, "POST", fullUrl, path.getPost());
+                addOperation(operationsAndResponses, "POST", url, path.getPost());
             }
 
             if (path.getDelete() != null) {
-                addOperation(operationsAndResponses, "DELETE", fullUrl, path.getDelete());
+                addOperation(operationsAndResponses, "DELETE", url, path.getDelete());
             }
         });
 
@@ -126,5 +106,27 @@ public class OpenApiSpec {
     private static void addOperation(Map<OpenApiOperation, Set<String>> operationsAndResponses, String method, String fullUrl, Operation operation) {
         Set<String> responseCodes = operation.getResponses() == null ? Collections.emptySet() : operation.getResponses().keySet();
         operationsAndResponses.put(new OpenApiOperation(method, fullUrl), responseCodes);
+    }
+
+    private OpenAPI createOpenAPI() {
+        if (!isSpecDefined) {
+            return null;
+        }
+
+        ParseOptions parseOptions = new ParseOptions();
+        parseOptions.setResolve(true);
+
+        SwaggerParseResult swaggerParseResult = new OpenAPIParser().readLocation(
+                specUrl, Collections.emptyList(), parseOptions);
+
+        OpenAPI openAPI = swaggerParseResult.getOpenAPI();
+
+        if (openAPI == null) {
+            throw new IllegalArgumentException(
+                    format("Unable to load API descriptor from provided %s:\n\t%s",
+                            specUrl, swaggerParseResult.getMessages().toString().replace("\n", "\n\t")));
+        }
+
+        return openAPI;
     }
 }

--- a/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpec.java
+++ b/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpec.java
@@ -123,10 +123,19 @@ public class OpenApiSpec {
 
         if (openAPI == null) {
             throw new IllegalArgumentException(
-                    format("Unable to load API descriptor from provided %s:\n\t%s",
-                            specUrl, swaggerParseResult.getMessages().toString().replace("\n", "\n\t")));
+                    format("Unable to load API descriptor from provided %s: %s", specUrl,
+                            parseResultMessage(swaggerParseResult)));
         }
 
         return openAPI;
+    }
+
+    private static String parseResultMessage(SwaggerParseResult swaggerParseResult) {
+        List<String> messages = swaggerParseResult.getMessages();
+        if (messages == null) {
+            return "check if it is accessible";
+        }
+
+        return "\n    " + String.join("\n    ", messages);
     }
 }

--- a/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpecConfig.java
+++ b/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpecConfig.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2020 webtau maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -23,7 +24,6 @@ import org.testingisdocumenting.webtau.cfg.WebTauConfigHandler;
 import java.util.stream.Stream;
 
 import static org.testingisdocumenting.webtau.cfg.ConfigValue.declare;
-import static org.testingisdocumenting.webtau.cfg.WebTauConfig.getCfg;
 
 public class OpenApiSpecConfig implements WebTauConfigHandler {
     static final ConfigValue specUrl = declare("openApiSpecUrl",
@@ -32,21 +32,32 @@ public class OpenApiSpecConfig implements WebTauConfigHandler {
     static final ConfigValue ignoreAdditionalProperties = declare("openApiIgnoreAdditionalProperties",
             "ignore additional OpenAPI properties ", () -> false);
 
-    private static String fullPath;
+    private static String fullPathOrUrl;
 
-    static String specFullPath() {
-        return fullPath;
+    static String getSpecFullPathOrUrl() {
+        return fullPathOrUrl;
     }
 
     @Override
     public void onAfterCreate(WebTauConfig cfg) {
-        fullPath = specUrl.getAsString().isEmpty() ? "" :
-                cfg.getWorkingDir().resolve(specUrl.getAsString()).toString();
+        fullPathOrUrl = resolveFullPathOrUrl(cfg, specUrl.getAsString());
         OpenApi.reset();
     }
 
     @Override
     public Stream<ConfigValue> additionalConfigValues() {
         return Stream.of(specUrl, ignoreAdditionalProperties);
+    }
+
+    private static String resolveFullPathOrUrl(WebTauConfig cfg, String configValue) {
+        if (configValue.isEmpty()) {
+            return "";
+        }
+
+        if (configValue.startsWith("http:") || configValue.startsWith("https:")) {
+            return configValue;
+        }
+
+        return cfg.getWorkingDir().resolve(specUrl.getAsString()).toString();
     }
 }

--- a/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpecValidator.java
+++ b/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpecValidator.java
@@ -1,4 +1,5 @@
 /*
+ * Copyright 2020 webtau maintainers
  * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -16,7 +17,7 @@
 
 package org.testingisdocumenting.webtau.openapi;
 
-import com.atlassian.oai.validator.SwaggerRequestResponseValidator;
+import com.atlassian.oai.validator.OpenApiInteractionValidator;
 import com.atlassian.oai.validator.model.SimpleRequest;
 import com.atlassian.oai.validator.model.SimpleResponse;
 import com.atlassian.oai.validator.report.LevelResolver;
@@ -27,21 +28,21 @@ import org.testingisdocumenting.webtau.http.validation.HttpValidationResult;
 
 import java.util.Optional;
 
+import static java.util.stream.Collectors.joining;
 import static org.testingisdocumenting.webtau.utils.UrlUtils.extractPath;
 import static org.testingisdocumenting.webtau.utils.UrlUtils.extractQueryParams;
-import static java.util.stream.Collectors.joining;
 
 public class OpenApiSpecValidator {
-    private final SwaggerRequestResponseValidator openApiValidator;
+    private final OpenApiInteractionValidator openApiValidator;
     private final OpenApiSpec openAPISpec;
 
     public OpenApiSpecValidator(OpenApiSpec openApiSpec, OpenApiValidationConfig validationConfig) {
         this.openAPISpec = openApiSpec;
         this.openApiValidator = openApiSpec.isSpecDefined() ?
-                SwaggerRequestResponseValidator
-                        .createFor(openApiSpec.getSpecUrl())
+                OpenApiInteractionValidator
+                        .createForSpecificationUrl(openApiSpec.getSpecUrl())
                         .withLevelResolver(createLevelResolver(validationConfig))
-                        .build():
+                        .build() :
                 null;
     }
 
@@ -51,7 +52,7 @@ public class OpenApiSpecValidator {
 
     public void validateApiSpec(HttpValidationResult result, ValidationMode validationMode) {
         Optional<OpenApiOperation> apiOperation = openAPISpec.findApiOperation(result.getRequestMethod(), result.getFullUrl());
-        if (! apiOperation.isPresent()) {
+        if (!apiOperation.isPresent()) {
             ConsoleOutputs.out(Color.YELLOW, "Path, ", result.getFullUrl(), " not found in OpenAPI spec");
             return;
         }
@@ -60,12 +61,13 @@ public class OpenApiSpecValidator {
         SimpleResponse response = buildResponse(result);
 
         ValidationReport validationReport = validate(validationMode, request, response);
-        validationReport.getMessages().forEach(message -> result.addMismatch("API spec validation failure: " + message.toString()));
+        validationReport.getMessages().forEach(message ->
+                result.addMismatch("API spec validation failure: " + renderMessage(message)));
 
         if (!validationReport.getMessages().isEmpty()) {
             throw new AssertionError("schema is not valid:\n" + validationReport
                     .getMessages().stream()
-                    .map(Object::toString)
+                    .map(OpenApiSpecValidator::renderMessage)
                     .collect(joining("\n")));
         }
     }
@@ -84,9 +86,17 @@ public class OpenApiSpecValidator {
         }
     }
 
+    private static String renderMessage(ValidationReport.Message message) {
+        return message.getContext().map(c ->
+                (c.getRequestPath().isPresent() && c.getRequestMethod().isPresent() ?
+                        (c.getRequestMethod().get() + " " + c.getRequestPath().get() + ": ") : "") +
+                        message.getMessage()).orElse("");
+    }
+
     private SimpleResponse buildResponse(HttpValidationResult result) {
         return SimpleResponse.Builder
                 .status(result.getResponseStatusCode())
+                .withContentType(result.getResponseType())
                 .withBody(result.getResponseTextContent())
                 .build();
     }
@@ -109,6 +119,7 @@ public class OpenApiSpecValidator {
 
     private LevelResolver createLevelResolver(OpenApiValidationConfig validationConfig) {
         LevelResolver.Builder builder = LevelResolver.create();
+
         if (validationConfig.isIgnoreAdditionalProperties()) {
             builder.withLevel("validation.schema.additionalProperties", ValidationReport.Level.IGNORE);
         }

--- a/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpecValidator.java
+++ b/webtau-open-api/src/main/java/org/testingisdocumenting/webtau/openapi/OpenApiSpecValidator.java
@@ -94,9 +94,14 @@ public class OpenApiSpecValidator {
     }
 
     private SimpleResponse buildResponse(HttpValidationResult result) {
-        return SimpleResponse.Builder
-                .status(result.getResponseStatusCode())
-                .withContentType(result.getResponseType())
+        SimpleResponse.Builder builder = SimpleResponse.Builder
+                .status(result.getResponseStatusCode());
+
+        if (!result.getResponseType().isEmpty()) {
+            builder.withContentType(result.getResponseType());
+        }
+
+        return builder
                 .withBody(result.getResponseTextContent())
                 .build();
     }

--- a/webtau-open-api/src/test/groovy/org/testingisdocumenting/webtau/openapi/OpenApiResponseValidatorTest.groovy
+++ b/webtau-open-api/src/test/groovy/org/testingisdocumenting/webtau/openapi/OpenApiResponseValidatorTest.groovy
@@ -63,9 +63,9 @@ class OpenApiResponseValidatorTest implements StepReporter {
         HttpValidationResult validationResult = createValidationResult()
 
         def expectedError = "schema is not valid:\n" +
-                "ERROR - No request body is expected for GET on path '/customer/{id}'.: []\n" +
-                "ERROR - Object instance has properties which are not allowed by the schema: [\"key\"]: []\n" +
-                "ERROR - Object has missing required properties ([\"mandatoryField\"]): []"
+                "GET /customer/2: No request body is expected but one was found.\n" +
+                "GET /customer/2: Object instance has properties which are not allowed by the schema: [\"key\"]\n" +
+                "GET /customer/2: Object has missing required properties ([\"mandatoryField\"])"
 
         code {
             HttpValidationHandlers.validate(validationResult)
@@ -93,8 +93,8 @@ class OpenApiResponseValidatorTest implements StepReporter {
         HttpValidationResult validationResult = createValidationResult()
 
         def expectedError = "schema is not valid:\n" +
-                "ERROR - Object instance has properties which are not allowed by the schema: [\"key\"]: []\n" +
-                "ERROR - Object has missing required properties ([\"mandatoryField\"]): []"
+                "GET /customer/2: Object instance has properties which are not allowed by the schema: [\"key\"]\n" +
+                "GET /customer/2: Object has missing required properties ([\"mandatoryField\"])"
 
         code {
             OpenApi.responseOnlyValidation {
@@ -111,7 +111,7 @@ class OpenApiResponseValidatorTest implements StepReporter {
         HttpValidationResult validationResult = createValidationResult()
 
         def expectedError = "schema is not valid:\n" +
-                "ERROR - No request body is expected for GET on path '/customer/{id}'.: []"
+                "GET /customer/2: No request body is expected but one was found."
         code {
             OpenApi.requestOnlyValidation() {
                 HttpValidationHandlers.validate(validationResult)
@@ -133,6 +133,7 @@ class OpenApiResponseValidatorTest implements StepReporter {
     private static HttpResponse response() {
         def response = new HttpResponse()
         response.setTextContent('{"key": "value"}')
+        response.setContentType('application/json')
         response.setStatusCode(200)
         response
     }

--- a/webtau-open-api/src/test/groovy/org/testingisdocumenting/webtau/openapi/OpenApiSpecValidatorTest.groovy
+++ b/webtau-open-api/src/test/groovy/org/testingisdocumenting/webtau/openapi/OpenApiSpecValidatorTest.groovy
@@ -81,6 +81,7 @@ class OpenApiSpecValidatorTest {
     static HttpResponse ok(String body) {
         def response = new HttpResponse()
         response.setTextContent(body)
+        response.setContentType("application/json")
         response.setStatusCode(200)
 
         return response

--- a/webtau-open-api/src/test/resources/test-spec.json
+++ b/webtau-open-api/src/test/resources/test-spec.json
@@ -37,14 +37,31 @@
       }
     },
     "/customer/{id}": {
+      "parameters": [
+        {"in": "path", "name": "id", "type": "string"}
+      ],
       "get": {
         "summary": "get customer",
         "operationId": "getCustomer",
         "responses": {
           "200": {
-            "description": "success",
+            "description": "success get response",
             "schema": {
-              "$ref": "#/definitions/TestResponse"
+              "type": "object",
+              "required": [
+                "mandatoryField"
+              ],
+              "properties": {
+                "mandatoryField": {
+                  "type": "string"
+                },
+                "optionalField": {
+                  "type": "string"
+                },
+                "intField": {
+                  "type": "integer"
+                }
+              }
             }
           },
           "4XX": {

--- a/webtau-testapp/pom.xml
+++ b/webtau-testapp/pom.xml
@@ -27,7 +27,6 @@
     <properties>
         <spring.boot.version>1.5.7.RELEASE</spring.boot.version>
         <hsqldb.version>2.4.0</hsqldb.version>
-        <springfox.version>2.8.0</springfox.version>
         <jaxb.version>2.3.0</jaxb.version>
     </properties>
 
@@ -94,18 +93,6 @@
         </dependency>
 
         <dependency>
-            <groupId>io.springfox</groupId>
-            <artifactId>springfox-swagger2</artifactId>
-            <version>${springfox.version}</version>
-        </dependency>
-
-        <dependency>
-            <groupId>io.springfox</groupId>
-            <artifactId>springfox-data-rest</artifactId>
-            <version>${springfox.version}</version>
-        </dependency>
-
-        <dependency>
             <groupId>javax.xml.bind</groupId>
             <artifactId>jaxb-api</artifactId>
             <version>${jaxb.version}</version>
@@ -121,6 +108,12 @@
             <groupId>com.sun.xml.bind</groupId>
             <artifactId>jaxb-impl</artifactId>
             <version>${jaxb.version}</version>
+        </dependency>
+
+        <dependency>
+            <groupId>org.springdoc</groupId>
+            <artifactId>springdoc-openapi-ui</artifactId>
+            <version>1.4.8</version>
         </dependency>
 
         <dependency>
@@ -142,7 +135,6 @@
                             <findAllDuplicates>true</findAllDuplicates>
                             <ignoreWhenIdentical>true</ignoreWhenIdentical>
                             <ignoreClasses>
-                                <!--tomcat embedded ships javax annotations internally-->
                                 <ignoreClass>javax.persistence.*</ignoreClass>
                             </ignoreClasses>
                         </banDuplicateClasses>

--- a/webtau-testapp/pom.xml
+++ b/webtau-testapp/pom.xml
@@ -28,6 +28,7 @@
         <spring.boot.version>1.5.7.RELEASE</spring.boot.version>
         <hsqldb.version>2.4.0</hsqldb.version>
         <jaxb.version>2.3.0</jaxb.version>
+        <enforcer.skip>true</enforcer.skip>
     </properties>
 
     <artifactId>webtau-testapp</artifactId>
@@ -122,25 +123,4 @@
             <version>1.1.1</version>
         </dependency>
     </dependencies>
-
-    <build>
-        <plugins>
-            <plugin>
-                <groupId>org.apache.maven.plugins</groupId>
-                <artifactId>maven-enforcer-plugin</artifactId>
-                <configuration>
-                    <rules>
-                        <dependencyConvergence />
-                        <banDuplicateClasses>
-                            <findAllDuplicates>true</findAllDuplicates>
-                            <ignoreWhenIdentical>true</ignoreWhenIdentical>
-                            <ignoreClasses>
-                                <ignoreClass>javax.persistence.*</ignoreClass>
-                            </ignoreClasses>
-                        </banDuplicateClasses>
-                    </rules>
-                </configuration>
-            </plugin>
-        </plugins>
-    </build>
 </project>

--- a/webtau-testapp/src/main/java/com/example/demo/springboot/app/SpringBootDemoApp.java
+++ b/webtau-testapp/src/main/java/com/example/demo/springboot/app/SpringBootDemoApp.java
@@ -1,32 +1,30 @@
+/*
+ * Copyright 2020 webtau maintainers
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package com.example.demo.springboot.app;
 
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.boot.autoconfigure.hateoas.HypermediaAutoConfiguration;
-import org.springframework.context.annotation.Bean;
-import org.springframework.context.annotation.Import;
 import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
-import springfox.documentation.builders.PathSelectors;
-import springfox.documentation.builders.RequestHandlerSelectors;
-import springfox.documentation.spi.DocumentationType;
-import springfox.documentation.spring.web.plugins.Docket;
-import springfox.documentation.swagger2.annotations.EnableSwagger2;
 
 @SpringBootApplication(exclude = HypermediaAutoConfiguration.class)
 @EnableJpaRepositories
-@EnableSwagger2
-@Import({springfox.documentation.spring.data.rest.configuration.SpringDataRestConfiguration.class})
 public class SpringBootDemoApp {
     public static void main(String[] args) {
         SpringApplication.run(SpringBootDemoApp.class, args);
-    }
-
-    @Bean
-    public Docket api() {
-        return new Docket(DocumentationType.SWAGGER_2)
-                .select()
-                .apis(RequestHandlerSelectors.any())
-                .paths(PathSelectors.any())
-                .build();
     }
 }

